### ISSUE TITLE
dials.index with real_space_grid_search but no unit_cell gives a traceback instead of printing a user-comprehensible error

### ISSUE THREAD

**Issue by benjaminhwilliams** [dials/dials#976]
If we adhere to the principle that the user should never see a traceback, then this is broken:
`dials.index indexing.method=real_space_grid_search` requires the user to suggest unit cell parameters with `indexing.known_symmetry.unit_cell`.  Providing no unit cell parameters should result in a printed warning and exit.  Instead, at present, we print the traceback included below.

In addition, the traceback itself doesn't clearly point to the source of the problem.  I think we need
1. a printed warning and exit if calling `dials.index` with insufficient information from the command line,
2. a more meaningful error thrown by something in `dials.algorithms.indexing`, for anyone building something that uses the functionality of dials indexing indirectly (\<cough\> screen19 \<cough\>).

I can easily fix the former, but I'm afraid I find `dials.algorithms.indexing` to be rather opaque and confusing and I can find no guide to its structure, so I'm all at sea on the best way to execute the latter.

```
Traceback (most recent call last):
  File "/home/ekm22040/DIALS/build/../modules/dials/command_line/index.py", line 276, in <module>
    run()
  File "/home/ekm22040/DIALS/build/../modules/dials/command_line/index.py", line 255, in run
    experiments, reflections, params
  File "/home/ekm22040/DIALS/build/../modules/dials/command_line/index.py", line 180, in index
    known_crystal_models=known_crystal_models,
  File "/home/ekm22040/DIALS/build/../modules/dials/command_line/index.py", line 118, in _index_experiments
    params=params,
  File "/home/ekm22040/DIALS/modules/dials/algorithms/indexing/indexer.py", line 375, in from_parameters
    reflections, experiments, params, known_crystal_models
  File "/home/ekm22040/DIALS/modules/dials/algorithms/indexing/known_orientation.py", line 12, in __init__
    super(IndexerKnownOrientation, self).__init__(reflections, experiments, params)
  File "/home/ekm22040/DIALS/modules/dials/algorithms/indexing/indexer.py", line 358, in __init__
    self.setup_indexing()
  File "/home/ekm22040/DIALS/modules/dials/algorithms/indexing/indexer.py", line 488, in setup_indexing
    self.find_max_cell()
  File "/home/ekm22040/DIALS/modules/dials/algorithms/indexing/indexer.py", line 880, in find_max_cell
    overlaps_border=params.overlaps_border,
  File "/home/ekm22040/DIALS/modules/dials/algorithms/indexing/max_cell.py", line 29, in find_max_cell
    ice_sel = ice_rings_selection(reflections)
  File "/home/ekm22040/DIALS/modules/dials/algorithms/spot_finding/per_image_analysis.py", line 509, in ice_rings_selection
    unit_cell, space_group, flex.min(d_spacings), width
  File "/home/ekm22040/DIALS/modules/dials/algorithms/integration/filtering.py", line 50, in __init__
    assert d_min > 0
AssertionError
```

**Comment by rjgildea** [comment]
The above traceback doesn't make sense as it refers to the known_orientation indexer, not real_space_grid_search. This is the traceback I get:
```
$ dials.index $(libtbx.find_in_repositories dials_regression)/indexing_test_data/i04_weak_data/{experiments_import.json,full.pickle} indexing.method=real_space_grid_search
DIALS (2018) Acta Cryst. D74, 85-97. https://doi.org/10.1107/S2059798317017235
DIALS 2.dev.1018-gf5e586018
The following parameters have been modified:

indexing {
  method = fft3d fft1d *real_space_grid_search low_res_spot_match
}
input {
  experiments = /dls/mx-scratch/rjgildea/software/cctbx/modules/dials_regression/indexing_test_data/i04_weak_data/experiments_import.json
  reflections = /dls/mx-scratch/rjgildea/software/cctbx/modules/dials_regression/indexing_test_data/i04_weak_data/full.pickle
}

Found max_cell: 199.2 Angstrom
Indexing from 116082 reflections
Traceback (most recent call last):
  File "/dls/mx-scratch/rjgildea/software/cctbx/build/../modules/dials/command_line/index.py", line 276, in <module>
    run()
  File "/dls/mx-scratch/rjgildea/software/cctbx/build/../modules/dials/command_line/index.py", line 255, in run
    experiments, reflections, params
  File "/dls/mx-scratch/rjgildea/software/cctbx/build/../modules/dials/command_line/index.py", line 180, in index
    known_crystal_models=known_crystal_models,
  File "/dls/mx-scratch/rjgildea/software/cctbx/build/../modules/dials/command_line/index.py", line 120, in _index_experiments
    idxr.index()
  File "/dls/mx-scratch/rjgildea/software/cctbx/modules/dials/algorithms/indexing/indexer.py", line 539, in index
    experiments.extend(self.find_lattices())
  File "/dls/mx-scratch/rjgildea/software/cctbx/modules/dials/algorithms/indexing/lattice_search/__init__.py", line 139, in find_lattices
    self.candidate_crystal_models = self.find_candidate_crystal_models()
  File "/dls/mx-scratch/rjgildea/software/cctbx/modules/dials/algorithms/indexing/lattice_search/__init__.py", line 349, in find_candidate_crystal_models
    self.find_candidate_basis_vectors()
  File "/dls/mx-scratch/rjgildea/software/cctbx/modules/dials/algorithms/indexing/lattice_search/__init__.py", line 337, in find_candidate_basis_vectors
    reflections["rlp"]
  File "/dls/mx-scratch/rjgildea/software/cctbx/modules/dials/algorithms/indexing/basis_vector_search/real_space_grid_search.py", line 114, in find_basis_vectors
    vectors, weights = self.score_vectors(reciprocal_lattice_vectors)
  File "/dls/mx-scratch/rjgildea/software/cctbx/modules/dials/algorithms/indexing/basis_vector_search/real_space_grid_search.py", line 98, in score_vectors
    for i, v in enumerate(self.search_vectors):
  File "/dls/mx-scratch/rjgildea/software/cctbx/modules/dials/algorithms/indexing/basis_vector_search/real_space_grid_search.py", line 66, in search_vectors
    unique_cell_dimensions = set(self._target_unit_cell.parameters()[:3])
AttributeError: 'NoneType' object has no attribute 'parameters'
```

**Comment by benjaminhwilliams** [comment]
Thanks @rjgildea.

Ah, apologies, I copied the wrong traceback — I was cooking at the same time.  After I found the problem described above, I started playing around.  The traceback above is what you get if you call `dials.index` on the output of a previous successful `dials.index` call without providing a unit cell, i.e.
```dials.index indexed.{expt,refl}```.

I'm not sure if that's also worth fixing or whether we regard it out of the scope of normal use.
